Cython/Compiler/ExprNodes.py

```diff
@@ -6761,10 +6761,10 @@ def analyse_types(self, env):
             for arg in self.keyword_args
         ]
 
-        if len(args) == 1 and args[0].type is dict_type:
+        if len(args) == 1 and args[0].type is dict_type and args[0].is_name and args[0].entry.is_arg:
```

Comment:
I'm trying to work out if this optimization is _ever_ the right thing to do. I don't `entry.is_arg` makes it any more valid.

It might be OK if args[0] is `DictNode` and `ComprehensionNode`?

But I think you should just delete this whole `if` block.

Reply:
I think this optimization is wrong: see https://stackoverflow.com/a/45883925/ - [the specification](https://docs.python.org/3/reference/compound_stmts.html#function-definitions) says that keyword arguments are copied into a new mapping.

Therefore I think the test should be deleted and replaced with a another test showing that modifying keyword arguments in a function does not modify them outside the function.

The other option is that the optimization could be cleverer - within `wrapper` `kwargs` is never looked at again so the optimization is safe there. I don't think Cython has the information to choose that though.

Reply:
Simple is beauty, I'll delete it.